File: 2020/Day1/golang/day1v2_part2.go
package main
	
import (
    "bufio"
    "fmt"
    "os"
    "strconv"
    "flag"
)

func readLines(path string) ([]int, error) {
    file, err := os.Open(path)
    if err != nil {
        return nil, err
    }
    defer file.Close()

    var lines []int
    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
        convertedLine, _ := strconv.Atoi(scanner.Text())
        lines = append(lines, convertedLine)
    }
    return lines, scanner.Err()
}

func check(e error) {
    if e != nil {
        panic(e)
    }
}

func getSumPair(elements []int, search int, posMap map[int]int, exclude int) (int, int) {
    var num1 int = -1
    var num2 int = -1
    for i, element := range elements {
        if(i == exclude) { continue };
        num2 = search - element
        val, ok := posMap[num2]
        if ok && val != exclude {
            num1 = element
            break
        }
    }

    return num1, num2
}

func main() {
    var search int
    flag.IntVar(&search ,"search", 2020, "search sum integer.")
    flag.Parse()

	lines, _ := readLines("../input.txt")

    var posMap map[int]int
    posMap = make(map[int]int)
    for i, element := range lines {
        posMap[element] = i
    }

    var num1 int = -1
    var num2 int = -1
    var num3 int = -1
    for i, element := range lines {
        subSearch := search - element
        num1, num2 = getSumPair(lines, subSearch, posMap, i)
        if (num1 != -1 && num2 != -1){
            num3 = element
            break
        }
    }

    if (num3 != -1){
        fmt.Printf("Magic Numbers: %d + %d + %d = %d.\n", num1, num2, num3, num1+num2+num3)
        fmt.Printf("The solution is: %d x %d x %d = %d.\n", num1, num2, num3, num1*num2*num3)
    } else {
        fmt.Println("No valid Tris found.")
    }

}

File: 2020/Day1/python/day1v2_part2.py
import os

def read_input(filename):
    return open(filename).readlines()

def get_sum_pair(elements, search, pos_map) -> tuple:
    num1 = None
    for element in elements:
        num2 = search - element
        if pos_map.get(num2):
            num1 = element
            break
    
    return num1, num2

if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(description='Process some integers.')
    parser.add_argument('search', metavar='search', type=int, help='The sum integer to find')
    args = parser.parse_args()

    input_lines = read_input(os.path.join(os.path.dirname(__file__),"../input.txt"))
    elements = [ int(x.strip()) for x in input_lines ]

    pos_map = dict(zip(elements, range(0, len(elements))))

    num3 = None
    for i, element in enumerate(elements):
        sub_search = args.search - element
        sub_elements = elements[:i] + elements[i+1:]
        sub_pos_map = pos_map.copy()
        _ = sub_pos_map.pop(element)

        num1, num2 = get_sum_pair(sub_elements, sub_search, sub_pos_map)
        if num1 and num2:
            num3 = element
            break
    
    if num3:
        print(f"Magic Numbers: {num1} + {num2} + {num3} = {num1+num2+num3}.")
        print(f"The solution is: {num1} x {num2} x {num3} = {num1*num2*num3}.")
    else:
        print("No valid tris found.")
